### DISCUSSION TITLE
React in footnotes?

### DISCUSSION THREAD

**Discussion by mosnicholas** [remarkjs/remark#1239] [Q&A] [answered]
I've written a custom plugin that converts uuids found in my app with a react button. The goal is pretty simple: it should extract UUIDs and replace them with a button that can visually render those DB references. See the screenshot.

<img width="544" alt="Screenshot 2023-10-06 at 5 39 07 PM" src="https://github.com/remarkjs/remark/assets/3606778/c4bcb9cb-c236-4d57-bd0b-105703847420">

The plugin is defined like so:

```
import reactStringReplace from 'react-string-replace';
import { visit } from 'unist-util-visit';

import InlineSourcesButton from 'components/chatbot/ChatThread/InlineSourcesButton';

const uuidRegex =
  /(\b[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}\b,?\s*)+/gi;

const remarkCitationPlugin = () => {
  return (tree) => {
    visit(tree, 'text', (node) => {
      const matches = node.value.match(uuidRegex);
      if (matches) {
        let mIndex = 0;
        node.type = 'jsx';
        node.value = reactStringReplace(node.value, uuidRegex, (_, i) => {
          const uuids = matches[mIndex]
            .split(/,\s*/)
            ?.filter(Boolean)
            ?.map((uuid: string) => uuid.trim());

          mIndex++;
          return <InlineSourcesButton key={i} uuids={uuids} />;
        });
      }
    });
  };
};
```

The issue I'm having is that using remark-gfm results in the following footnote structure. I'm guessing this has something to do with the order in which plugins are applied and how remark-gfm renders footnotes (as HTML vs as JSX elements).
<img width="281" alt="Screenshot 2023-10-06 at 5 40 30 PM" src="https://github.com/remarkjs/remark/assets/3606778/db04a762-714f-4be8-ac75-77401bec9a24">

Is there a way to control how the output is rendered in remark-gfm footnotes?  My wrapper on ReactMarkdown is super simple:

```
<ReactMarkdown remarkPlugins={[remarkCitationPlugin, remarkGfm]}>{text}</ReactMarkdown
```

Thank you!

**Comment by ChristianMurphy** [comment, marked as answer]
Welcome @mosnicholas!
Sorry you ran into a spot of trouble.

The issue is a misunderstanding of what `mdast` and `remarkPlugins` are.
Start by reviewing https://github.com/remarkjs/react-markdown#architecture and https://unifiedjs.com/learn/
Note that, `remark` plugins and `mdast` do not contain react components (which is why you are seeing `[object Object]`).

Your end goal is to add a new content type which doesn't currently exist in `mdast` a source button, there are some pointers on how to do this in https://github.com/orgs/remarkjs/discussions/867

**Reply by mosnicholas** [reply]
First off, thank you for the quick response!

it seems like it's time to read.. my rough understanding browsing your links is that i'm trying to render React directly inline, whereas what I should be doing is transforming my UUIDs to a new "component" type that I then teach react markdown how to render via the `components` prop. Is that roughly correct?

**Reply by ChristianMurphy** [reply]
Roughly yes, to me more precise:
Add it into a Markdown AST (`mdast`) node with some HTML AST (`hast`) metadata attached, so your custom react component can be attached to that HTML tag through the `components` prop.

**Reply by mosnicholas** [reply]
Wow this is so much simpler. Thank you @ChristianMurphy ! Just to confirm, the discussion you linked had a lot of metadata attached on each regex find. This seems to be working too, and is a lot simpler:
```
const uuidRegex =
  /(\b[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}\b,?\s*)+/gi;

const remarkCitationPlugin = () => {
  return (tree) => {
    visit(tree, 'text', (node) => {
      const matches = node.value.match(uuidRegex);
      if (matches) {
        node.type = 'sourceUUIDs';
        node.data = {
          hName: 'sourceUUIDs',
          hProperties: {
            uuids: matches?.filter(Boolean)?.map((uuid: string) => uuid.trim()),
          },
        };
      }
    });
  };
};
```

Is there anything wrong with this?

**Reply by ChristianMurphy** [reply]
I'm glad you have a working version @mosnicholas!
A few thoughts:
1. If you need to support matches, where there is text before or after the uuid, the code would need to be updated. Currently it replaces all of the text surrounding the uuid.
2. Custom elements in HTML generally should be lower case and contain a dash https://developer.mozilla.org/en-US/docs/Web/API/Web_Components/Using_custom_elements. Camel case naming is non-standard and unspecified, it may work now, but is not guaranteed to work in the future. Considering using a more standard custom element name.

**Reply by mosnicholas** [reply]
Epic, thank you @ChristianMurphy ! I’m guessing we need to include the position start and end logic to inform the parser where the uuids are in the text?

two quick follow ups:
1. Any advice on handling new “components” in typescript?
2. I’m seeing the uuids passed back into my component as text as opposed to as a string array. Is that to be expected?

Thank you!

**Reply by ChristianMurphy** [reply]
>  I’m guessing we need to include the position start and end logic to inform the parser where the uuids are in the text?

ASTs work off of nodes.
To have a text, UUID, then text again.
You'd need to access the parent of the text, and splice three child nodes where one text node used to be.
Parent is the third parameter passed into the visitor callback function https://github.com/syntax-tree/unist-util-visit#visitor

> two quick follow ups

1. yes, see https://github.com/remarkjs/react-markdown/issues/622#issuecomment-888490496
2. yes, arrays exist only in JS, HTML attributes are text. Consider serializing with a separator character like `,` or `|` to make it easier to do `string.split` later and retrieve the array.

**Reply by mosnicholas** [reply]
FWIW, the typescript you linked to did not work, but this did:

```
declare namespace JSX {
  interface IntrinsicElements {
    'element-name': { property: boolean };
  }
}
```

Sourced from [here](https://www.typescriptlang.org/docs/handbook/jsx.html#intrinsic-elements)

**Reply by mosnicholas** [reply]
Thank you so much for your help @ChristianMurphy !! 🙏

**Reply by mosnicholas** [reply]
Last question - is there a way to replace the components inline vs. on new lines? taking a look at remark-gfm to see how it's being done there, but curious if there's a simple way to accomplish that?

<img width="270" alt="Screenshot 2023-10-08 at 12 30 48 PM" src="https://github.com/remarkjs/remark/assets/3606778/d67a1551-0c78-454a-9f5f-3374e14d47a1">

**Reply by mosnicholas** [reply]
Oh I just found the [mdast-util-find-and-replace](https://github.com/syntax-tree/mdast-util-find-and-replace#replace) function. This simplifies the code a substantial bit, and fixes the issue ^

**Reply by JanRuettinger** [reply]
@mosnicholas  Could you post a short snippet on how you solved your issue with mdast-util-find-and-replace?

**Reply by mosnicholas** [reply]
@JanRuettinger - much simpler.

```
const remarkCitationPlugin = () => {
  return (tree: Nodes) => {
    const newTree = findAndReplace(tree, [
      uuidRegex,
      (match: string) => {
        const uuids = match
          .split(',')
          .map((uuid) => uuid.trim())
          .join(',');

        return {
          value: uuids,
          type: 'text',
          data: {
            hName: NODE_TYPE,
            hProperties: {
              uuids,
            },
          },
        };
      },
    ]);
    return newTree;
  };
};
```
